below/tc/src/types.rs

```diff
@@ -0,0 +1,182 @@
+pub enum QDisc {
+    FqCodel(FqCodelQDisc),
+}
+
+impl QDisc {
+    fn new(kind: &str, opts: Vec<TcOption>) -> Option<Self> {
+        if kind == FQ_CODEL {
+            let mut fq_codel = FqCodelQDisc::default();
+            for opt in opts {
+                match opt {
+                    TcOption::FqCodel(fq_codel_opt) => {
+                        match fq_codel_opt {
+                            TcQdiscFqCodelOption::Target(target) => fq_codel.target = target,
+                            TcQdiscFqCodelOption::Limit(limit) => fq_codel.limit = limit,
+                            TcQdiscFqCodelOption::Interval(interval) => fq_codel.interval = interval,
+                            TcQdiscFqCodelOption::Ecn(ecn) => fq_codel.ecn = ecn,
+                            TcQdiscFqCodelOption::Flows(flows) => fq_codel.flows = flows,
+                            TcQdiscFqCodelOption::Quantum(quantum) => fq_codel.quantum = quantum,
+                            TcQdiscFqCodelOption::CeThreshold(ce_threshold) => fq_codel.ce_threshold = ce_threshold,
+                            TcQdiscFqCodelOption::DropBatchSize(drop_batch_size) => fq_codel.drop_batch_size = drop_batch_size,
+                            TcQdiscFqCodelOption::MemoryLimit(memory_limit) => fq_codel.memory_limit = memory_limit,
+                            _ => {}
+                        }
+                    }
+                    _ => {}
+                }
+            }
+            return Some(Self::FqCodel(fq_codel));
+        }
+        None
+    }
+}
+
+#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub enum XStats {
+    FqCodel(FqCodelXStats),
+}
+
+#[derive(Default, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct FqCodelQDisc {
+    pub target: u32,
+    pub limit: u32,
+    pub interval: u32,
+    pub ecn: u32,
+    pub flows: u32,
+    pub quantum: u32,
+    pub ce_threshold: u32,
+    pub drop_batch_size: u32,
+    pub memory_limit: u32,
+}
+
+#[derive(Default, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct FqCodelXStats {
+    pub maxpacket: u32,
+    pub drop_overlimit: u32,
+    pub ecn_mark: u32,
+    pub new_flow_count: u32,
+    pub new_flows_len: u32,
+    pub old_flows_len: u32,
+    pub ce_mark: u32,
+    pub memory_usage: u32,
+    pub drop_overmemory: u32,
+}
+
+impl FqCodelXStats {
+    pub fn new(xstats: &TcFqCodelXstats) -> Self {
+        match xstats {
+            TcFqCodelXstats::Qdisc(qdisc) => Self {
+                maxpacket: qdisc.maxpacket,
+                drop_overlimit: qdisc.drop_overlimit,
+                ecn_mark: qdisc.ecn_mark,
+                new_flow_count: qdisc.new_flow_count,
+                new_flows_len: qdisc.new_flows_len,
+                old_flows_len: qdisc.old_flows_len,
+                ce_mark: qdisc.ce_mark,
+                memory_usage: qdisc.memory_usage,
+                drop_overmemory: qdisc.drop_overmemory,
+            },
+            _ => Self::default(),
```

Comment:
question for the second, makes sense to only deal with `TcFqCodelQdStats` but say in the future we want to expose class stats `TcFqCodelClStats` as well, then we'll have to change it back to something similar as above.
I wanted for `FqCodelXStats` to represent a combined struct of `Qdisc` and `Class`, if it's a `Qdisc` the fields relevant to `Class` would be `None` and vice-versa.

Or would it better to have another struct identify a classful FqCodel xstats if and when we need to? Then it would make sense to implement your second suggestion.

Reply:
> Or would it better to have another struct identify a classful FqCodel xstats if and when we need to?

I don't have a strong preference for either - I'm not familiar with this stuff enough to have a strong opinion :), but my feeling is it might be better to split it so each struct has clear meaning? I'll leave the decision up to you.

Right now I don't think it's taking either approach though. We'd want to make the fields Option<> if we are to futureproof it for overloading with `TcFqCodelClStats`.

Reply:
have addressed this in commit https://github.com/facebookincubator/below/pull/8210/commits/97ac6bf79e9c22db05d2443c781988d75d6ec54c
create an enum wrapper which can be extended with `TcFqCodelClStats` when we decide to add classful qdiscs support